code.js

```diff
@@ -1,3 +1,53 @@
 function tsp_hk(distance_matrix) {
-    return -1;
+    const numCities = distance_matrix.length; //num Cities = n  
+    if (numCities <= 1) return 0; //Case for 0 or 1
+    
+    const memo = new Map();
+
+    //Helper function to find shortest distance with Help-Karp
+    //FST = Find Shortest Tour  
+    //'unvisitedCities' is used as a bitmask where it shows if it is visited (1) or not visited (0)
+    //currentCity is the current city that is being processed
+    function FST(unvisitedCities, currentCity) { 
+        const citiesKey = `${unvisitedCities}-${currentCity}`; // Key to store the subset result 
+
+        //Check if subset has been solved
+        if (memo.has(citiesKey)) {
+            return memo.get(citiesKey);
+        }
+
+        //Base case: If all cities have been visited. return distance to last.
+        //'unvisitedCities === (1 << numCities) - 1' means all bits are set (all cities have been visited) 
```

Comment:
Still don't understand this -- `unvisitedCities` should be the list of unvisited cities, right? Then the condition is just `unvisitedCities.length === 0`, which is much easier to understand.

Reply:
Or is this actually the visited cities?

Reply:
Yes, it suppose to be visitedcities. In my original code I had the currentCity set as the visitedCity but changed it over to currentCity and never changed the unvisited cities I thought of it as it was marking the cities that were visited in a separate set and then was still looking through the unvisited cities. When I wrote the bit-mask up I said, wait a minute this is marking the visited cities rather than the unvisited cities. I have changed the name to visited. That was my mistake not updating the code after I had changed the name of the variables. Sorry if my email said something else. I was at a meet and I did not have a chance to look at what I had orginally put for this code because it took me like three days to do it.